# Building a Master Dictionary of Valid Words

We'll build a dictionary of words and phrases that could be valid clues for Codenames.  There are two primary categories:
* words such as cat, boat and king
* proper nouns which may be more than one word such as names (George Washington) or titles (Animal Farm)

In [1]:
import json
import re
import pandas as pd

# Words

We'll words from the [ConceptNet](http://conceptnet.io/) database.  ConceptNet is a network which connects words, phrases and concepts together.  This will be very useful to us later on, but for now we'll just use it for building our dictionary.

First, load the data which was found here: https://github.com/commonsense/conceptnet5/wiki/Downloads

In [2]:
df = pd.read_csv('dictionary/conceptnet-assertions-5.5.5.csv.gz', sep='\t', header=None, names=['uri', 'relation', 'from', 'to', 'json'])
df.head()

uri    relation  \
0      /a/[/r/Antonym/,/c/ab/агыруа/n/,/c/ab/аҧсуа/]  /r/Antonym   
1  /a/[/r/Antonym/,/c/adx/thəχ_kwo/a/,/c/adx/ʂap_...  /r/Antonym   
2    /a/[/r/Antonym/,/c/adx/tok_po/a/,/c/adx/ʂa_wə/]  /r/Antonym   
3    /a/[/r/Antonym/,/c/adx/ʂa_wə/a/,/c/adx/tok_po/]  /r/Antonym   
4  /a/[/r/Antonym/,/c/adx/ʂap_wə/a/,/c/adx/thəχ_k...  /r/Antonym   

                from               to  \
0     /c/ab/агыруа/n      /c/ab/аҧсуа   
1  /c/adx/thəχ_kwo/a    /c/adx/ʂap_wə   
2    /c/adx/tok_po/a     /c/adx/ʂa_wə   
3     /c/adx/ʂa_wə/a    /c/adx/tok_po   
4    /c/adx/ʂap_wə/a  /c/adx/thəχ_kwo   

                                                json  
0  {"dataset": "/d/wiktionary/en", "license": "cc...  
1  {"dataset": "/d/wiktionary/fr", "license": "cc...  
2  {"dataset": "/d/wiktionary/fr", "license": "cc...  
3  {"dataset": "/d/wiktionary/fr", "license": "cc...  
4  {"dataset": "/d/wiktionary/fr", "license": "cc...

`from` and `to` are the words that are connected and `relation` is how they are connected.  For now we'll just consider the `from` column. We can use the metadata on the word to only keep English words (`en`), and then we'll strip off the metadata.

In [3]:
df = df[['from', 'json']].drop_duplicates(subset='from')
df = df[(df['from'].str.extract(r'/./([^/]*)/.*') == 'en')]
df['from'] = df['from'].str.extract(r'/./[^/]*/([^/]*).*')
df.head()

c:\Users\Jeremy\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  
c:\Users\Jeremy\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  This is separate from the ipykernel package so we can avoid doing imports until


from                                               json
7469              0  {"dataset": "/d/wiktionary/fr", "license": "cc...
7470  12_hour_clock  {"dataset": "/d/wiktionary/en", "license": "cc...
7471  24_hour_clock  {"dataset": "/d/wiktionary/en", "license": "cc...
7472              5  {"dataset": "/d/wiktionary/en", "license": "cc...
7473            a.c  {"dataset": "/d/wiktionary/fr", "license": "cc...

There are several different [knowledge sources](https://github.com/commonsense/conceptnet5/wiki/Knowledge-sources) for ConceptNet.  The two we'll keep are the English Wikitionary and Verbosity.  

In [4]:
# Keep: wiktionary and verbosity
df = df[
    df['json'].str.contains('verbosity') | df['json'].str.contains('/d/wiktionary/en')
]
df.head()

from                                               json
7470  12_hour_clock  {"dataset": "/d/wiktionary/en", "license": "cc...
7471  24_hour_clock  {"dataset": "/d/wiktionary/en", "license": "cc...
7472              5  {"dataset": "/d/wiktionary/en", "license": "cc...
7474            a.m  {"dataset": "/d/wiktionary/en", "license": "cc...
7477       ab_extra  {"dataset": "/d/wiktionary/en", "license": "cc...

Finally we'll remove any multi-word rows:

In [5]:
words = df['from']
words = words[~words.str.contains('_')].drop_duplicates()
print(words.head())
print()
print("Number of words:", len(words))

7472            5
7474          a.m
7479    abactinal
7480      abandon
7482     abapical
Name: from, dtype: object

Number of words: 492938


And we're left with 492938 "words".

# Proper Nouns

Next, we'll build a list of valid proper nouns.  ConceptNet does include some from [DBPedia](http://wiki.dbpedia.org/), but we'll go straight to the source to get a more complete list.  DBPedia is a project to extract structured data from Wikipedia.  They release many different datasets here: http://wiki.dbpedia.org/downloads-2016-10

The one that we'll use right now is the English Language instance type data.  Fer each article on Wikipedia, it will be labeled with what it is such as a person, place or book.  We can filter out only the objects of types which could have proper nouns.

First, we read in the data.  It comes as a `ttl` file which I found easiest to just parse line-by-line.

In [6]:
import re

rows = []
with open('dictionary/instance_types_en.ttl', 'r', encoding="utf8") as f:
    f.readline()
    for l in f.readlines():
        try:
            split = l.split(' ')
            rows.append((split[0].split('/')[-1], split[2]))
        except:
            print(l)

The two columns of interest are the object name and the type:

In [7]:
df = pd.DataFrame.from_records(rows[:-1], columns=['object', 'type'])
df.head(10)

object                                               type
0             Anarchism>              <http://www.w3.org/2002/07/owl#Thing>
1              Achilles>              <http://www.w3.org/2002/07/owl#Thing>
2                Autism>              <http://dbpedia.org/ontology/Disease>
3               Alabama>  <http://dbpedia.org/ontology/AdministrativeReg...
4       Abraham_Lincoln>         <http://dbpedia.org/ontology/OfficeHolder>
5    Abraham_Lincoln__1>           <http://dbpedia.org/ontology/TimePeriod>
6    Abraham_Lincoln__2>           <http://dbpedia.org/ontology/TimePeriod>
7    Abraham_Lincoln__3>           <http://dbpedia.org/ontology/TimePeriod>
8  An_American_in_Paris>              <http://www.w3.org/2002/07/owl#Thing>
9       Animalia_(book)>                 <http://dbpedia.org/ontology/Book>

After cleaning it up we're left with:

In [8]:
df = df[df['type'].str.contains('dbpedia.org/ontology')].copy()
df['object'] = df['object'].str.strip(r'[<>]').str.split('__', expand=True)[0].str.replace('_', ' ').str.replace(r'\(.*\)', '').str.strip()
df = df.drop_duplicates()
df['type'] = df['type'].str.strip(r'[<>]').str.split('/', expand=True)[4]
df['object'] = df['object'].str.replace(r'\s+', ' ')
df['object'] = df['object'].str.replace('%22', '"')
df['object'] = df['object'].str.replace('%3F', '?')
df = df[~(df['object'] == '')].copy()

df.head(10)

object                  type
2            Autism               Disease
3           Alabama  AdministrativeRegion
4   Abraham Lincoln          OfficeHolder
5   Abraham Lincoln            TimePeriod
9          Animalia                  Book
10   Academy Awards                 Award
11          Actrius                  Film
13       Allan Dwan                Person
14       Allan Dwan        PersonFunction
15     Alain Connes             Scientist

There are hundreds of nuanced instance types.  To figure out which might be "proper noun" types, I took a partially manual approach.  I generated stats on the number of words, number of capitalized words, etc. and then by hand sorted through them to decide which ones seemed to consist of proper nouns and which didn't.

In [9]:
def countUpper(x):
    words = x.split(' ')
    return len([w for w in words if w[0].isupper()]) - 1

def countWords(x):
    return len(x.split(' '))

def countNumbers(x):
    return len(re.findall(r'[0-9]', x))

df['num_upper'] = df['object'].apply(countUpper)
df['num_words'] = df['object'].apply(countWords)
df['num_numbers'] = df['object'].apply(countNumbers)
df['percent_upper'] = df['num_upper']/(df['num_words'] - 1)

df[df['num_words'] > 1].groupby('type').mean().sort_values('percent_upper', ascending=False).to_csv('dictionary/type_stats.csv')

df[df['num_words'] > 1].groupby('type').mean().sort_values('percent_upper', ascending=False)

num_upper  num_words  num_numbers  \
type                                                               
BoxingLeague                    3.000000   4.000000     0.000000   
ChemicalElement                 1.000000   2.000000     0.000000   
HorseTrainer                    1.580247   2.580247     0.000000   
VideogamesLeague                2.000000   3.000000     0.000000   
LacrossePlayer                  1.023148   2.023148     0.000000   
Guitarist                       1.006849   2.006849     0.000000   
SpeedwayLeague                  2.153846   3.153846     0.000000   
SnookerChamp                    1.000000   2.000000     0.000000   
VoiceActor                      1.001873   2.001873     0.000000   
BowlingLeague                   2.000000   3.000000     0.000000   
Bay                             1.222222   2.222222     0.000000   
Constellation                   1.000000   2.000000     0.000000   
CricketLeague                   2.000000   3.000000     0.200000   
Presenter                       1.035398   2.035398     0.000000   
MixedMartialArtsLeague          2.000000   3.000000     0.000000   
CurlingLeague                   2.666667   3.666667     0.000000   
SumoWrestler                    1.031780   2.031780     0.000000   
Stream                          1.155556   2.155556     0.000000   
VolleyballCoach                 1.065789   2.065789     0.000000   
SpaceShuttle                    2.000000   3.000000     0.200000   
Spacecraft                      1.500000   2.500000     0.000000   
WaterwayTunnel                  1.291667   2.291667     0.000000   
AmericanFootballPlayer          1.029693   2.030831     0.000000   
NascarDriver                    1.075322   2.077304     0.000000   
AustralianRulesFootballPlayer   1.015773   2.017141     0.000000   
GridironFootballPlayer          1.024160   2.025832     0.000152   
IceHockeyPlayer                 1.011910   2.013970     0.000000   
CollegeCoach                    1.154626   2.157302     0.000000   
Congressman                     1.648328   2.650760     0.000000   
GaelicGamesPlayer               1.031133   2.034304     0.000000   
...                                  ...        ...          ...   
Archaea                         0.033333   2.044444     0.088889   
Fish                            0.019253   2.094876     0.000293   
Asteroid                        0.015442   2.019765     3.680667   
Brain                           0.017613   2.653620     0.166341   
Lymph                           0.012987   3.350649     0.000000   
Fern                            0.008463   2.032440     0.000000   
Bacteria                        0.008292   2.024876     0.008292   
Artery                          0.008850   3.365782     0.000000   
Plant                           0.005337   2.025709     0.000071   
Muscle                          0.003802   3.022814     0.000000   
Bone                            0.003145   2.748428     0.000000   
Conifer                         0.004680   2.023401     0.000000   
Mollusca                        0.003189   2.015112     0.000323   
Animal                          0.003181   2.030046     0.000000   
Crustacean                      0.003808   2.041889     0.000000   
Eukaryote                       0.002833   2.025496     0.000000   
Arachnid                        0.001921   2.022574     0.000000   
Insect                          0.000907   2.007026     0.000124   
Fungus                          0.000342   2.036942     0.000000   
Cycad                           0.000000   2.000000     0.000000   
Embryology                      0.000000   2.165517     0.000000   
ClubMoss                        0.000000   2.024691     0.000000   
Moss                            0.000000   2.000000     0.000000   
Nerve                           0.000000   3.411043     0.000000   
Gnetophytes                     0.000000   2.000000     0.000000   
Ligament                        0.000000   3.027322     0.000000   
Vein               

After sorting through them manually I ended up with the 336 types in [`proper_noun_types.csv`](proper_noun_types.csv).

In [10]:
proper_noun_types = pd.read_csv('proper_noun_types.csv')
proper_noun_types

proper_nouns
0                        CricketLeague
1                         SpaceShuttle
2                            Guitarist
3                         HorseTrainer
4                       LacrossePlayer
5               MixedMartialArtsLeague
6                        CurlingLeague
7                               Stream
8                         SumoWrestler
9                       SpeedwayLeague
10                    VideogamesLeague
11                          VoiceActor
12                     VolleyballCoach
13                      WaterwayTunnel
14                          Spacecraft
15                           Presenter
16                        SnookerChamp
17                       Constellation
18                     ChemicalElement
19                        BoxingLeague
20                       BowlingLeague
21                                 Bay
22              AmericanFootballPlayer
23                        NascarDriver
24       AustralianRulesFootballPlayer
25              GridironFootballPlayer
26                     IceHockeyPlayer
27                        CollegeCoach
28                         Congressman
29                   GaelicGamesPlayer
..                                 ...
306                               Poem
307                            Country
308                       HandballTeam
309                       MilitaryUnit
310                        RailwayLine
311                        CricketTeam
312                         Locomotive
313                ProgrammingLanguage
314  SupremeCourtOfTheUnitedStatesCase
315          ClassicalMusicComposition
316                              Grape
317                 MythologicalFigure
318                               Road
319                     RailwayStation
320                   MilitaryConflict
321                               Star
322                            Fashion
323                 BiologicalDatabase
324                              Novel
325                        EthnicGroup
326                            Station
327                             Weapon
328                         Skyscraper
329                       SpaceStation
330                            Surname
331                              Train
332                ArtificialSatellite
333                             Galaxy
334                             Planet
335                           Asteroid

[336 rows x 1 columns]

There's still a little bit more filtering and cleaning to go.  First, I filterd down to the proper noun types.  And I noticed there were many `TimePeriod`s that were names, so I removed the ones with commas (more on that it a second).  

In [11]:
proper_nouns = df[
    df['type'].isin(proper_noun_types['proper_nouns']) &
    ~((df['type'] == 'TimePeriod') & (df['object'].str.contains(',')))
]
proper_nouns

object                  type  \
3                                         Alabama  AdministrativeRegion   
4                                 Abraham Lincoln          OfficeHolder   
5                                 Abraham Lincoln            TimePeriod   
9                                        Animalia                  Book   
10                                 Academy Awards                 Award   
11                                        Actrius                  Film   
13                                     Allan Dwan                Person   
14                                     Allan Dwan        PersonFunction   
15                                   Alain Connes             Scientist   
16                                      Aristotle           Philosopher   
17       Academy Award for Best Production Design                 Award   
19                                       Ayn Rand                Writer   
23                                        Algeria               Country   
29                                   Andre Agassi          TennisPlayer   
30                                    Animal Farm                  Book   
32                                        Andorra               Country   
33                                         Alaska  AdministrativeRegion   
41                                  Aldous Huxley                Writer   
43                          America the Beautiful                  Song   
44                          America the Beautiful                 Sound   
48          American National Standards Institute          Organisation   
50                              A Modest Proposal                  Book   
51                                       Apollo 8   ArtificialSatellite   
52                                      Apollo 11   ArtificialSatellite   
53                                      Apollo 11                 Sound   
58                               Andrei Tarkovsky                Person   
59                               Andrei Tarkovsky        PersonFunction   
60                               Andrei Tarkovsky           WrittenWork   
63                                           Asia             Continent   
64                                 Atlantic Ocean           BodyOfWater   
...                                           ...                   ...   
5150392         International Prize in Statistics                 Award   
5150393                            Karel San Juan          OfficeHolder   
5150394                         Eleousa, Ioannina  AdministrativeRegion   
5150395                                     Wingo               Airline   
5150396                             Magae Station               Station   
5150397                   The Gift of the Morning                  Film   
5150399                            Arturo Falconi                Person   
5150400                            Arturo Falconi        PersonFunction   
5150402                             Ander Bardají          SoccerPlayer   
5150403                             Ander Bardají         CareerStation   
5150408         Commercial Capital Training Group               Company   
5150409                  Karuppu Raja Vellai Raja                  Film   
5150410                   Malachi Mitchell-Thomas       MotorcycleRider   
5150411                    Rio Urubu State Forest         ProtectedArea   
5150412                            James Anderson          SoccerPlayer   
5150413                            James Anderson         CareerStation   
5150414                               Jane McLean                Person   
5150415                            Zakir Kathrada             Cricketer   
5150416                             Blayde Capell             Cricketer   
5150417                              Colin Buchan          SoccerPlayer   
5150418                              Colin Buchan         CareerStation   
5150422                               Hugh Butler          SoccerPla

Then I wanted to clean up article titles which included commas such as place names (Rome, Ohio) or people with titles.  So instead of having an entry in the dictionary that was  "Rome, Ohio" there would just be "Rome."

Again, through a manual process I identified the following types to do the split on.

In [12]:
split = [
    'Settlement', 'Village', 'City', 'Town', 'AdministrativeRegion', 'HistoricBuilding', 
    'Noble', 'School', 'OfficeHolder', 'Building', 'Royalty', 'HistoricPlace', 'University', 
    'Person', 'PersonFunction', 'Place', 'ReligiousBuilding', 'MilitaryPerson', 'Baronet',
    'MilitaryUnit', 'Politician', 'Road', 'Station', 'Island', 'RailwayLine', 'Museum',
    'Park', 'Venue', 'MemberOfParliament', 'Hospital', 'Prison', 'Mountain', 'River',
    'MilitaryStructure', 'Bridge', 'Hotel', 'Lake', 'Airport', 'College', 'Stadium',
    'SiteOfSpecialScientificInterest', 'Saint', 'Monarch', 'WorldHeritageSite', 
    'Governor', 'ShoppingMall', 'ChristianBishop', 'Monument', 'CricketGround'
]

proper_nouns.loc[
    proper_nouns['object'].str.contains(',') & proper_nouns['type'].isin(split),
    'object'
] = proper_nouns[
    proper_nouns['object'].str.contains(',') & proper_nouns['type'].isin(split)
]['object'].str.split(',', expand=True)[0]

proper_nouns

c:\Users\Jeremy\Anaconda3\lib\site-packages\pandas\core\indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


object                  type  \
3                                         Alabama  AdministrativeRegion   
4                                 Abraham Lincoln          OfficeHolder   
5                                 Abraham Lincoln            TimePeriod   
9                                        Animalia                  Book   
10                                 Academy Awards                 Award   
11                                        Actrius                  Film   
13                                     Allan Dwan                Person   
14                                     Allan Dwan        PersonFunction   
15                                   Alain Connes             Scientist   
16                                      Aristotle           Philosopher   
17       Academy Award for Best Production Design                 Award   
19                                       Ayn Rand                Writer   
23                                        Algeria               Country   
29                                   Andre Agassi          TennisPlayer   
30                                    Animal Farm                  Book   
32                                        Andorra               Country   
33                                         Alaska  AdministrativeRegion   
41                                  Aldous Huxley                Writer   
43                          America the Beautiful                  Song   
44                          America the Beautiful                 Sound   
48          American National Standards Institute          Organisation   
50                              A Modest Proposal                  Book   
51                                       Apollo 8   ArtificialSatellite   
52                                      Apollo 11   ArtificialSatellite   
53                                      Apollo 11                 Sound   
58                               Andrei Tarkovsky                Person   
59                               Andrei Tarkovsky        PersonFunction   
60                               Andrei Tarkovsky           WrittenWork   
63                                           Asia             Continent   
64                                 Atlantic Ocean           BodyOfWater   
...                                           ...                   ...   
5150392         International Prize in Statistics                 Award   
5150393                            Karel San Juan          OfficeHolder   
5150394                                   Eleousa  AdministrativeRegion   
5150395                                     Wingo               Airline   
5150396                             Magae Station               Station   
5150397                   The Gift of the Morning                  Film   
5150399                            Arturo Falconi                Person   
5150400                            Arturo Falconi        PersonFunction   
5150402                             Ander Bardají          SoccerPlayer   
5150403                             Ander Bardají         CareerStation   
5150408         Commercial Capital Training Group               Company   
5150409                  Karuppu Raja Vellai Raja                  Film   
5150410                   Malachi Mitchell-Thomas       MotorcycleRider   
5150411                    Rio Urubu State Forest         ProtectedArea   
5150412                            James Anderson          SoccerPlayer   
5150413                            James Anderson         CareerStation   
5150414                               Jane McLean                Person   
5150415                            Zakir Kathrada             Cricketer   
5150416                             Blayde Capell             Cricketer   
5150417                              Colin Buchan          SoccerPlayer   
5150418                              Colin Buchan         CareerStation   
5150422                               Hugh Butler          SoccerPla

Then, to make the format match the words from ConceptNet, I made everything lowercase and replaced spaces with underscores.

In [13]:
proper_noun_words = proper_nouns['object'].str.lower().str.replace(' ', '_').drop_duplicates()
proper_noun_words = proper_noun_words[
    ~proper_noun_words.str.contains(r'^list_of')
]
proper_noun_words

3                                           alabama
4                                   abraham_lincoln
9                                          animalia
10                                   academy_awards
11                                          actrius
13                                       allan_dwan
15                                     alain_connes
16                                        aristotle
17         academy_award_for_best_production_design
19                                         ayn_rand
23                                          algeria
29                                     andre_agassi
30                                      animal_farm
32                                          andorra
33                                           alaska
41                                    aldous_huxley
43                            america_the_beautiful
48            american_national_standards_institute
50                                a_modest_proposal
51          

After all that, we'er left with a master dictionary of about 2.75 million words and proper nouns.

In [14]:
dic = dict([(w, 1) for w in (list(words) + list(proper_noun_words))])
len(dic)

2740624

In [15]:
import json
json.dump(dic, open('dictionary/words.json', 'w'))